### PR TITLE
Type refactor

### PR DESCRIPTION
**Pull request by marcandre**
This PR fixes two bugs:
- passing a bad `type`
- different behavior between `type: {...}` and `enum: {...}`

It also improve the style and simplifies the code.

### CODE REVIEW

File: lib/tlaw/params/type.rb [hunk 1]
```diff
@@ -30,28 +28,26 @@ def to_doc_type
       end
 
       def convert(value)
-        validate(value) && _convert(value)
+        if (err = validation_error(value))
+          fail Nonconvertible,
+               "#{self} can't convert  #{value.inspect}: #{err}"
+        end
+        _convert(value)
       end
```
Comment (zverok) [inline lib/tlaw/params/type.rb]: Why this change? I am not sure that `validate && convert` is a perfectly good code (what if `validate` decides value valid, but returns `nil`?), but the core of two methods named this way seems way more natural to me than "`validation_error`, which returns some text if there is an error".

File: spec/tlaw/param_spec.rb [hunk 2]
```diff
@@ -166,6 +166,12 @@ module Params
         end
       end
 
+      describe '#type' do
+        it 'raises an ArgumentError on bad argument' do
+          expect { described_class.new(:p, type: 'hello') }.to raise_error(ArgumentError)
```
Comment (zverok) [inline spec/tlaw/param_spec.rb]: Maybe add test for err message?..

File: lib/tlaw/params/type.rb (hunk 1 above)
Comment (marcandre) [reply lib/tlaw/params/type.rb]: My idea was to factorize code. The pattern `check_something or nonconvertible!(value, "not an instance of #{type}")` is there in a few places; personally I don't like it, but of course might be different for you , especially that you are the author.
Note that these are private classes and thus methods...

File: spec/tlaw/param_spec.rb (hunk 2 above)
Comment (marcandre) [reply spec/tlaw/param_spec.rb]: Sure, fixing...

Comment (marcandre) [reply spec/tlaw/param_spec.rb]: Done.
